helix-term/src/ui/editor.rs

```diff
@@ -144,19 +165,53 @@ impl EditorView {
                     theme,
                     &config.cursor_shape,
                 ),
-            ))
+            );
+            let focused_view_elements = Self::highlight_focused_view_elements(view, doc, theme);
+            if focused_view_elements.is_empty() {
+                Box::new(highlights)
+            } else {
+                Box::new(syntax::merge(highlights, focused_view_elements))
+            }
         } else {
             Box::new(highlights)
         };
 
-        Self::render_text_highlights(doc, view.offset, inner, surface, theme, highlights, &config);
-        Self::render_gutter(editor, doc, view, view.area, surface, theme, is_focused);
-        Self::render_rulers(editor, doc, view, inner, surface, theme);
+        Self::render_gutter(
+            editor,
+            doc,
+            view,
+            view.area,
+            theme,
+            is_focused,
+            &mut line_decorations,
+        );
 
         if is_focused {
-            Self::render_focused_view_elements(view, doc, inner, theme, surface);
+            let cursor = doc
+                .selection(view.id)
+                .primary()
+                .cursor(doc.text().slice(..));
+            // set the cursor_cache to out of view in case the position is not found
+            editor.cursor_cache.set(Some(None));
+            let update_cursor_cache =
+                |_: &mut TextRenderer, pos| editor.cursor_cache.set(Some(Some(pos)));
+            translated_positions.push((cursor, Box::new(update_cursor_cache)));
         }
 
+        // TODO allow smooth scrolling softwrap/virtual text
```

Comment:
Smooth scrolling of soft wrapped document text seems to work already, so maybe "soft wrapped virtual text" instead of "softwrap/virtual text" might be clearer.  Unless I'm misunderstanding the comment.  But the current phrasing makes it sound like smooth scrolling of soft wrapped text generally, as well as virtual text, doesn't work yet.

Reply:
this TODO is really old from my first PR #5008 and I just forgot to remove it.
Smooth scrolling always works (and is very much non-option as many functions miss-behave without it)